File: src-tauri/src/export/remap.rs
use crate::edit::clip::Clip;
use crate::edit::model::{Cut, Speed, Trim};

pub const FACTOR_MIN: f64 = 0.25;
pub const FACTOR_MAX: f64 = 8.0;

#[derive(Clone, Debug, PartialEq)]
pub struct Segment {
    pub clip_start: u32,
    pub clip_end: u32,
    pub factor: f64,
    pub out_start: f64,
    pub clip: usize,
}

#[derive(Clone, Debug)]
pub struct TimeMap {
    segs: Vec<Segment>,
    out_dur: f64,
    trim_in: u32,
    plain: bool,
}

fn merged_cuts(cuts: &[Cut], lo: u32, hi: u32) -> Vec<(u32, u32)> {
    let mut v: Vec<(u32, u32)> = cuts
        .iter()
        .map(|c| (c.start_ms.clamp(lo, hi), c.end_ms.clamp(lo, hi)))
        .filter(|(a, b)| a < b)
        .collect();
    v.sort_unstable();
    let mut out: Vec<(u32, u32)> = Vec::new();
    for (a, b) in v {
        match out.last_mut() {
            Some(last) if a <= last.1 => last.1 = last.1.max(b),
            _ => out.push((a, b)),
        }
    }
    out
}

fn clamped_spans(speed: &[Speed], lo: u32, hi: u32) -> Vec<(u32, u32, f64)> {
    let mut v: Vec<(u32, u32, f64)> = speed
        .iter()
        .map(|s| {
            (
                s.start_ms.clamp(lo, hi),
                s.end_ms.clamp(lo, hi),
                (s.factor as f64).clamp(FACTOR_MIN, FACTOR_MAX),
            )
        })
        .collect();
    v.sort_by(|a, b| a.0.cmp(&b.0));
    let mut out: Vec<(u32, u32, f64)> = Vec::new();
    for (a, b, f) in v {
        let a = out.last().map_or(a, |l| a.max(l.1));
        if a < b {
            out.push((a, b, f));
        }
    }
    out
}

fn clip_ranges(trim: &Trim, clips: &[Clip], full_dur_ms: u32) -> Vec<(usize, u32, u32)> {
    if clips.is_empty() {
        let (lo, hi) = trim.resolve(full_dur_ms);
        return vec![(0, lo, hi)];
    }
    clips
        .iter()
        .enumerate()
        .map(|(i, c)| {
            (
                i,
                c.src_in_ms.min(full_dur_ms),
                c.src_out_ms.min(full_dur_ms),
            )
        })
        .filter(|(_, a, b)| a < b)
        .collect()
}

impl TimeMap {
    pub fn build(
        trim: &Trim,
        cuts: &[Cut],
        speed: &[Speed],
        clips: &[Clip],
        full_dur_ms: u32,
    ) -> TimeMap {
        let (lo, hi) = trim.resolve(full_dur_ms);
        let ranges = clip_ranges(trim, clips, full_dur_ms);
        let mut segs: Vec<Segment> = Vec::new();
        let mut out_start = 0.0f64;
        let mut plain = ranges.len() == 1 && (ranges[0].1, ranges[0].2) == (lo, hi);
        for &(ci, rlo, rhi) in ranges.iter() {
            let cuts = merged_cuts(cuts, rlo, rhi);
            let spans = clamped_spans(speed, rlo, rhi);
            plain = plain && cuts.is_empty() && spans.is_empty();
            let mut kept: Vec<(u32, u32)> = Vec::new();
            let mut at = rlo;
            for (a, b) in &cuts {
                if at < *a {
                    kept.push((at, *a));
                }
                at = at.max(*b);
            }
            if at < rhi {
                kept.push((at, rhi));
            }
            for (keep_start, keep_end) in kept {
                let mut edges: Vec<u32> = vec![keep_start, keep_end];
                for (a, b, _) in &spans {
                    for e in [a, b] {
                        if *e > keep_start && *e < keep_end {
                            edges.push(*e);
                        }
                    }
                }
                edges.sort_unstable();
                edges.dedup();
                for w in edges.windows(2) {
                    let (seg_start, seg_end) = (w[0], w[1]);
                    let factor = spans
                        .iter()
                        .find(|(a, b, _)| *a <= seg_start && seg_end <= *b)
                        .map_or(1.0, |s| s.2);
                    segs.push(Segment {
                        clip_start: seg_start,
                        clip_end: seg_end,
                        factor,
                        out_start,
                        clip: ci,
                    });
                    out_start += (seg_end - seg_start) as f64 / factor;
                }
            }
        }
        TimeMap {
            segs,
            out_dur: out_start,
            trim_in: lo,
            plain,
        }
    }

    pub fn identity(full_dur_ms: u32) -> TimeMap {
        TimeMap::build(&Trim::default(), &[], &[], &[], full_dur_ms)
    }
    pub fn is_plain(&self) -> bool {
        self.plain
    }
    pub fn segments(&self) -> &[Segment] {
        &self.segs
    }
    pub fn out_dur_ms(&self) -> u32 {
        self.out_dur.round() as u32
    }

    pub fn out_of(&self, clip_ms: u32) -> u32 {
        if let Some(s) = self
            .segs
            .iter()
            .find(|s| s.clip_start <= clip_ms && clip_ms < s.clip_end)
        {
            return (s.out_start + (clip_ms - s.clip_start) as f64 / s.factor).round() as u32;
        }
        self.segs
            .iter()
            .filter(|s| s.clip_start > clip_ms)
            .min_by(|a, b| {
                (a.clip_start, a.out_start)
                    .partial_cmp(&(b.clip_start, b.out_start))
                    .unwrap()
            })
            .map_or(self.out_dur.round() as u32, |s| s.out_start.round() as u32)
    }

    pub fn clip_of(&self, out_ms: u32) -> u32 {
        let o = out_ms as f64;
        for s in &self.segs {
            let len = (s.clip_end - s.clip_start) as f64 / s.factor;
            if o < s.out_start + len {
                return (s.clip_start as f64 + (o - s.out_start) * s.factor).round() as u32;
            }
        }
        self.segs.last().map_or(self.trim_in, |s| s.clip_end)
    }

    fn seg_of_out(&self, out_ms: u32) -> Option<usize> {
        let o = out_ms as f64;
        self.segs
            .iter()
            .position(|s| o < s.out_start + (s.clip_end - s.clip_start) as f64 / s.factor)
    }

    pub fn crosses_boundary(&self, prev_out_ms: u32, out_ms: u32) -> bool {
        match (self.seg_of_out(prev_out_ms), self.seg_of_out(out_ms)) {
            (Some(a), Some(b)) if a < b => {
                (a..b).any(|i| self.segs[i].clip_end != self.segs[i + 1].clip_start)
            }
            _ => false,
        }
    }

    pub fn clip_out_ms(&self, clip: usize) -> u32 {
        self.segs
            .iter()
            .filter(|s| s.clip == clip)
            .map(|s| (s.clip_end - s.clip_start) as f64 / s.factor)
            .sum::<f64>()
            .round() as u32
    }

    pub fn frame_bounds(&self, i: usize, fps: u64) -> Option<(u64, u64)> {
        let s = self.segs.get(i)?;
        let (seg_start, seg_end) = (s.clip_start as u64, s.clip_end as u64);
        let k_start = if i == 0 {
            seg_start * fps / 1000
        } else {
            (seg_start * fps + 999) / 1000
        };
        let k_end = if i + 1 == self.segs.len() {
            seg_end * fps / 1000
        } else {
            ((seg_end * fps + 999) / 1000).checked_sub(1)?
        };
        (k_end >= k_start).then_some((k_start, k_end))
    }

    pub fn frame_plan(&self, fps: u64) -> Vec<u64> {
        let mut plan = Vec::new();
        for (i, s) in self.segs.iter().enumerate() {
            let Some((k_start, k_end)) = self.frame_bounds(i, fps) else {
                continue;
            };
            let n = ((k_end - k_start) as f64 / s.factor).floor() as u64 + 1;
            plan.extend((0..n).map(|j| k_start + (j as f64 * s.factor).floor() as u64));
        }
        plan
    }

    pub fn plan_boundaries(&self, fps: u64) -> Vec<usize> {
        let mut out = Vec::new();
        let (mut last_end, mut at) = (None::<u32>, 0usize);
        for (i, s) in self.segs.iter().enumerate() {
            let Some((k_start, k_end)) = self.frame_bounds(i, fps) else {
                continue;
            };
            if last_end.is_some_and(|e| e != s.clip_start) {
                out.push(at);
            }
            last_end = Some(s.clip_end);
            at += ((k_end - k_start) as f64 / s.factor).floor() as usize + 1;
        }
        out
    }
}

#[cfg(test)]
#[path = "remap_tests.rs"]
pub(crate) mod tests;

#[cfg(test)]
#[path = "remap_clips_tests.rs"]
pub(crate) mod clips_tests;

#[path = "remap_spans.rs"]
mod spans;
pub use spans::ClipSpan;

File: src/shared/math/remap.ts
import type { Clip, Cut, Speed, Trim } from "../edit";

export const FACTOR_MIN = 0.25;
export const FACTOR_MAX = 8;

export interface Segment {
  clipStart: number;
  clipEnd: number;
  factor: number;
  outStart: number;
  clip: number;
}

export interface TimeMap {
  readonly segments: Segment[];
  readonly outDur: number;
  readonly trimIn: number;
  readonly plain: boolean;
}

const clamp = (v: number, lo: number, hi: number) => Math.min(hi, Math.max(lo, v));

export function resolveTrim(trim: Trim, fullDurMs: number): [number, number] {
  const out = trim.out_ms === 0 ? fullDurMs : Math.min(trim.out_ms, fullDurMs);
  return [Math.min(trim.in_ms, out), out];
}

function mergedCuts(cuts: Cut[], lo: number, hi: number): [number, number][] {
  const v = cuts
    .map((c): [number, number] => [clamp(c.start_ms, lo, hi), clamp(c.end_ms, lo, hi)])
    .filter(([a, b]) => a < b);
  v.sort((x, y) => x[0] - y[0] || x[1] - y[1]);
  const out: [number, number][] = [];
  for (const [a, b] of v) {
    const last = out[out.length - 1];
    if (last && a <= last[1]) last[1] = Math.max(last[1], b);
    else out.push([a, b]);
  }
  return out;
}

function clampedSpans(speed: Speed[], lo: number, hi: number): [number, number, number][] {
  const v = speed.map((s): [number, number, number] => [
    clamp(s.start_ms, lo, hi),
    clamp(s.end_ms, lo, hi),
    clamp(s.factor, FACTOR_MIN, FACTOR_MAX),
  ]);
  v.sort((x, y) => x[0] - y[0]);
  const out: [number, number, number][] = [];
  for (const [a0, b, f] of v) {
    const last = out[out.length - 1];
    const a = last ? Math.max(a0, last[1]) : a0;
    if (a < b) out.push([a, b, f]);
  }
  return out;
}

function clipRanges(trim: Trim, clips: Clip[], fullDurMs: number): [number, number, number][] {
  if (clips.length === 0) {
    const [lo, hi] = resolveTrim(trim, fullDurMs);
    return [[0, lo, hi]];
  }
  return clips
    .map((c, i): [number, number, number] => [
      i,
      Math.min(c.src_in_ms, fullDurMs),
      Math.min(c.src_out_ms, fullDurMs),
    ])
    .filter(([, a, b]) => a < b);
}

export function buildTimeMap(
  trim: Trim,
  cuts: Cut[],
  speed: Speed[],
  clips: Clip[],
  fullDurMs: number,
): TimeMap {
  const [lo, hi] = resolveTrim(trim, fullDurMs);
  const ranges = clipRanges(trim, clips, fullDurMs);
  const segments: Segment[] = [];
  let outStart = 0;
  let plain = ranges.length === 1 && ranges[0][1] === lo && ranges[0][2] === hi;
  for (const [ci, rlo, rhi] of ranges) {
    const merged = mergedCuts(cuts, rlo, rhi);
    const spans = clampedSpans(speed, rlo, rhi);
    plain = plain && merged.length === 0 && spans.length === 0;
    const kept: [number, number][] = [];
    let at = rlo;
    for (const [a, b] of merged) {
      if (at < a) kept.push([at, a]);
      at = Math.max(at, b);
    }
    if (at < rhi) kept.push([at, rhi]);
    for (const [ks, ke] of kept) {
      const edges = [ks, ke];
      for (const [a, b] of spans) for (const e of [a, b]) if (e > ks && e < ke) edges.push(e);
      const uniq = [...new Set(edges)].sort((x, y) => x - y);
      for (let i = 0; i + 1 < uniq.length; i++) {
        const [cs, ce] = [uniq[i], uniq[i + 1]];
        const span = spans.find(([a, b]) => a <= cs && ce <= b);
        const factor = span ? span[2] : 1;
        segments.push({ clipStart: cs, clipEnd: ce, factor, outStart, clip: ci });
        outStart += (ce - cs) / factor;
      }
    }
  }
  return { segments, outDur: outStart, trimIn: lo, plain };
}

export const identityMap = (fullDurMs: number): TimeMap =>
  buildTimeMap({ in_ms: 0, out_ms: 0 }, [], [], [], fullDurMs);

export const outDurMs = (m: TimeMap): number => Math.round(m.outDur);

export function nextShown(m: TimeMap, clipMs: number): Segment | undefined {
  let next: Segment | undefined;
  for (const g of m.segments)
    if (
      g.clipStart > clipMs &&
      (!next ||
        g.clipStart < next.clipStart ||
        (g.clipStart === next.clipStart && g.outStart < next.outStart))
    )
      next = g;
  return next;
}

export function outOf(m: TimeMap, clipMs: number): number {
  const s = m.segments.find((g) => g.clipStart <= clipMs && clipMs < g.clipEnd);
  if (s) return Math.round(s.outStart + (clipMs - s.clipStart) / s.factor);
  const next = nextShown(m, clipMs);
  return Math.round(next ? next.outStart : m.outDur);
}

export function clipOf(m: TimeMap, outMs: number): number {
  for (const s of m.segments) {
    const len = (s.clipEnd - s.clipStart) / s.factor;
    if (outMs < s.outStart + len) return Math.round(s.clipStart + (outMs - s.outStart) * s.factor);
  }
  const last = m.segments[m.segments.length - 1];
  return last ? last.clipEnd : m.trimIn;
}

function segOfOut(m: TimeMap, outMs: number): number {
  return m.segments.findIndex((s) => outMs < s.outStart + (s.clipEnd - s.clipStart) / s.factor);
}

export function crossesBoundary(m: TimeMap, prevOutMs: number, outMs: number): boolean {
  const [a, b] = [segOfOut(m, prevOutMs), segOfOut(m, outMs)];
  if (a < 0 || b < 0 || a >= b) return false;
  for (let i = a; i < b; i++) if (m.segments[i].clipEnd !== m.segments[i + 1].clipStart) return true;
  return false;
}

export function clipOutMs(m: TimeMap, clip: number): number {
  return Math.round(
    m.segments
      .filter((s) => s.clip === clip)
      .reduce((acc, s) => acc + (s.clipEnd - s.clipStart) / s.factor, 0),
  );
}

export function factorAt(m: TimeMap, clipMs: number): number {
  const s = m.segments.find((g) => clipMs >= g.clipStart && clipMs < g.clipEnd);
  return s ? s.factor : 1;
}

export function frameBounds(m: TimeMap, i: number, fps: number): [number, number] | null {
  const s = m.segments[i];
  if (!s) return null;
  const kStart =
    i === 0 ? Math.floor((s.clipStart * fps) / 1000) : Math.floor((s.clipStart * fps + 999) / 1000);
  const kEnd =
    i + 1 === m.segments.length
      ? Math.floor((s.clipEnd * fps) / 1000)
      : Math.floor((s.clipEnd * fps + 999) / 1000) - 1;
  return kEnd >= kStart && kEnd >= 0 ? [kStart, kEnd] : null;
}

export function framePlan(m: TimeMap, fps: number): number[] {
  const plan: number[] = [];
  m.segments.forEach((s, i) => {
    const b = frameBounds(m, i, fps);
    if (!b) return;
    const n = Math.floor((b[1] - b[0]) / s.factor) + 1;
    for (let j = 0; j < n; j++) plan.push(b[0] + Math.floor(j * s.factor));
  });
  return plan;
}
